doc/rados/operations/monitoring-osd-pg.rst

```diff
@@ -109,9 +109,15 @@ requires three replicas of a placement group, CRUSH may assign them to
 ``osd.1``, ``osd.2`` and ``osd.3`` respectively. CRUSH actually seeks a
 pseudo-random placement that will take into account failure domains you set in
 your `CRUSH map`_, so you will rarely see placement groups assigned to nearest
-neighbor OSDs in a large cluster. We refer to the set of OSDs that should
-contain the replicas of a particular placement group as the **Acting Set**. In
-some cases, an OSD in the Acting Set is ``down`` or otherwise not able to
+neighbor OSDs in a large cluster.
+
+Ceph processes a client request using the **Acting Set**, which is the set of
```

Comment:
> Ceph processes a client request using the **Up Set**

i still think it's up set which handles the requests. for understanding up set, we need to introduce pg temp. see https://docs.ceph.com/en/latest/dev/peering/#concepts .

considering an acting set of [0,1,2] of some certain PG. since the the PG is active and clean, the up set is identical to the acting set.

if osd.0 cannot serve the request anymore, CRUSH might pick [3,1,2] as the new acting set. but since osd.3 does not contain the pg shard for serving the requests yet. it asks monitor for a "temp pg" of [1,3,2]. once monitor approves / commits this change. the new up set of this PG is changed to [1,3,2] and the acting set is still [3,1,2]. before osd.3 is backfilled, it's [1,3,2] who is serving the client requests. and once osd.3 is ready, both acting set and up set are [3,1,2].

Reply:
ok, sure i can add an explainatin for `pg_temp`, too.

about the up/acting - I take my understanding from the following observations:
* when an osd is down and out, and a pg becomes degraded+remapped, the sets are:
  * `up=[38, 36, 28, 17, 13, 39, 48, 10, 29, 5, 47]`
  * `acting=[36, 15, 28, 17, 13, 32, -1, 10, 29, 5, 20]`
  * -> 48 is not ready yet, waiting to be filled
* when I add `pg_upmap_item 11.ae8 22 50` to `up=acting=[110,22,190,210,89,310,270]`
  * `up=[110,50,190,210,89,310,270]`
  * `acting=[110,22,190,210,89,310,270]`
  * -> 50 is not ready yet and it is filled
  * meanwhile, 22 is still acting as it has the full data available

So `up` is the "future", `acting` the "current". of course, and there you have to help me if this is true, client requests can be handled by both. the up set will have less data available, so I assumed it's better to query the acting set.

I understand the [peering concept](https://docs.ceph.com/en/latest/dev/peering/#concepts) document the same way:
* Acting set: the ordered list of OSDs who are (...) responsible for a particular PG -> current osds, not future
* Primary: the (by convention first) member of the acting set, who is responsible for coordination peering, and is the only OSD that will accept client initiated writes to objects in a placement group -> acting set[0] = client initiated writes

So this would mean `acting set`=client requests, current responsible OSDs. `up set`= planned future location of the pg. And that's what I tried to clarify in my patch.

Reply:
searching for proof in the linux code, client requests are [processed](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ceph/osd_client.c?h=v5.8#n2358) [by the](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ceph/osd_client.c?h=v5.8#n1640) [acting](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ceph/osdmap.c?h=v5.8#n2691) [set indeed](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ceph/osd_client.c?h=v5.8#n1665).

Reply:
I don't think we should mention `pg_temp` here, this is an internal detail in my opinion.